filament/backend/src/opengl/OpenGLDriver.h

```diff
@@ -373,8 +386,16 @@ class OpenGLDriver final : public backend::DriverBase {
     void setExternalTexture(GLTexture* t, void* image);
 
     void whenGpuCommandsComplete(std::function<void()> fn) noexcept;
+    void whenFrameBegins(std::function<bool()> fn) noexcept;
+
     void executeGpuCommandsCompleteOps() noexcept;
-    std::vector<std::pair<GLsync, std::function<void(void)>>> mGpuCommandCompleteOps;
+    void executeFrameBeginsOps() noexcept;
+
+    std::vector<std::pair<GLsync, std::function<void()>>> mGpuCommandCompleteOps;
+    std::vector<std::function<bool()>> mFrameBeginsOps;
```

Comment:
There is a lot about these two vectors that would not be obvious to newcomers.  Here is my understanding:

- The purpose of these queues is to execute code on the main thread.
- Both of these task queues execute code at the beginning of the frame.
- Tasks that return a boolean can optionally persist.
- Tasks that are associated with a GLsync will only execute if the sync is satisfied.

Maybe this code could be generalized (and made easier to understand) if we unify the queues and design a small "Task" class that holds an optional GLsync handle.

At the very least it probably needs some comments.  :)

Reply:
all good points, I'll go over this in a follow-up PR